crypto-primitives/src/sponge/rescue/constraints.rs

```diff
@@ -0,0 +1,337 @@
+use crate::sponge::constraints::AbsorbGadget;
+use crate::sponge::constraints::{CryptographicSpongeVar, SpongeWithGadget};
+use crate::sponge::rescue::{RescueConfig, RescueSponge};
+use crate::sponge::DuplexSpongeMode;
+use ark_ff::PrimeField;
+use ark_r1cs_std::fields::fp::FpVar;
+use ark_r1cs_std::prelude::*;
+use ark_relations::gr1cs::{ConstraintSystemRef, SynthesisError};
+
+#[cfg(not(feature = "std"))]
+use ark_std::vec::Vec;
+
+pub const RESCUE_PREDICATE: &str = "Deg5-Mul";
```

Comment:
TODO: 

1. Rename `RESCUE_PREDICATE` to `DEG_5_MUL`
2. Add `DEG_3_MUL` and `DEG_7_MUL` predicates.
3. Add helper function to register the predicate in the constraint system
4. Add helper function which does
```
fn has_custom_gate(alpha: u64) → Option<&'static str> {
    match alpha {
        3 => Some(DEG_3_MUL),
        5 => Some(DEG_5_MUL),
        7 => Some(DEG_7_MUL),
        _ => None,
    }
}
```